compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/XCLBinGen.cpp

```diff
@@ -828,45 +828,27 @@ LogicalResult xilinx::aie2xclbin(MLIRContext *ctx, ModuleOp moduleOp,
   PassManager pm(ctx, moduleOp.getOperationName());
   applyConfigToPassManager(TK, pm);
 
-  if (TK.Verbose) {
-    llvm::outs() << "Running: ";
-    pm.printAsTextualPipeline(llvm::outs());
-    llvm::outs() << "\n";
-  }
-
-  if (failed(pm.run(moduleOp)))
-    return moduleOp.emitOpError("AIE lowering pipline failed");
-
-  std::regex target_regex("AIE.?");
-  if (!std::regex_search(TK.TargetArch, target_regex))
-    return moduleOp.emitOpError()
-           << "Unexpected target architecture: " << TK.TargetArch;
-
   // generateNPUInstructions
-  {
-    PassManager pm(ctx, moduleOp.getOperationName());
-    applyConfigToPassManager(TK, pm);
-
-    pm.addNestedPass<AIE::DeviceOp>(
-        mlir::iree_compiler::AMDAIE::createAIEDmaToNpuPass());
-    ModuleOp copy = moduleOp.clone();
-    if (failed(pm.run(copy)))
-      return moduleOp.emitOpError("NPU Instruction pipeline failed");
-
-    std::string errorMessage;
-    auto output = openOutputFile(OutputNPU, &errorMessage);
-    if (!output) {
-      llvm::errs() << errorMessage << "\n";
-      return moduleOp.emitOpError("");
-    }
+  pm.addNestedPass<AIE::DeviceOp>(
+      mlir::iree_compiler::AMDAIE::createAIEDmaToNpuPass());
+  if (failed(pm.run(moduleOp)))
+    return moduleOp.emitOpError(": NPU Instruction pipeline failed");
 
-    if (failed(
-            mlir::iree_compiler::AMDAIE::AIETranslateToNPU(copy, output->os())))
-      return moduleOp.emitOpError("NPU Instruction translation failed");
+  // TODO(max): should be using UI32 resource or something like that...
+  ArrayRef<int32_t> signedNpuInstructionsAttr =
+      cast<DenseI32ArrayAttr>(
+          (*moduleOp.getOps<xilinx::AIE::DeviceOp>().begin())
+              ->getAttr("npu_instructions"))
+          .asArrayRef();
+  std::vector<uint32_t> unsignedNpuInstructions(
+      signedNpuInstructionsAttr.begin(), signedNpuInstructionsAttr.end());
 
-    output->keep();
-    copy->erase();
-  }
+  std::string errorMessage;
+  auto output = openOutputFile(OutputNPU, &errorMessage);
+  if (!output) return moduleOp.emitOpError(errorMessage);
```

Comment:
Is the plan that this write-to-file will be removed too, eventually?

Reply:
Yes. Looking ahead, we will do this the same exact way to that GPU dialects do this - the runtime calls themselves are emitted into the IR (as `func.call` -> `llvm.call` -> C FFI calls) that is then executed for real real at runtime. So we will just emit a call somewhere (my rough/hazy vision is that IREE `hal` dialect/dispatch region will do this) that passes this sequence of instructions directly to some equivalent of `xrt::bo(npu_instructions, XCL_BO_CACHEABLE)`. @nirvedhmeshram I believe has been "hands-on" keyboard for the XRT connection so he can comment/paint a picture here.

Reply:
That's a clear explanation, thanks